gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/VectoredIOImpl.java

```diff
@@ -0,0 +1,289 @@
+  private final URI gcsPath;
+  private final FileInfo fileInfo;
+  private final GoogleCloudStorageReadOptions channelReadOptions;
+  private final VectoredReadOptions vectoredReadOptions;
+  private final GoogleCloudStorageFileSystem gcsFs;
+  private ExecutorService boundedThreadPool;
+
+  public VectoredIOImpl(
+      GoogleCloudStorageFileSystem gcsFs,
+      URI gcsPath,
+      FileInfo fileInfo,
+      VectoredReadOptions vectoredReadOptions) {
+    this.gcsFs = gcsFs;
+    this.gcsPath = gcsPath;
+    this.fileInfo = fileInfo;
+    this.vectoredReadOptions = vectoredReadOptions;
+    this.channelReadOptions =
+        channelReadOptions(gcsFs.getOptions().getCloudStorageOptions().getReadChannelOptions());
+    this.boundedThreadPool =
+        Executors.newFixedThreadPool(this.vectoredReadOptions.getReadThreads());
+  }
+
+  /**
+   * Reads data from Google Cloud Storage using vectored I/O operations.
+   *
+   * @param ranges List of file ranges to read.
+   * @param allocate Function to allocate ByteBuffer for reading.
+   * @throws IOException if an I/O error occurs.
+   */
+  public void readVectored(List<? extends FileRange> ranges, IntFunction<ByteBuffer> allocate)
+      throws IOException {
+    List<? extends FileRange> sortedRanges = validateNonOverlappingAndReturnSortedRanges(ranges);
+    for (FileRange range : ranges) {
+      validateRangeRequest(range);
+      CompletableFuture<ByteBuffer> result = new CompletableFuture<>();
+      range.setData(result);
+    }
+
+    if (shouldMergeRanges(ranges)) {
+      // case when ranges are not merged
+      for (FileRange range : sortedRanges) {
+        boundedThreadPool.submit(() -> readSingleRange(range, allocate));
+      }
+    } else {
+      // case where ranges can be merged
+      List<CombinedFileRange> combinedFileRanges =
+          mergeSortedRanges(
+              sortedRanges,
+              1,
+              vectoredReadOptions.getMinSeekVectoredReadSize(),
+              vectoredReadOptions.getMergeRangeMaxSize());
+      for (CombinedFileRange combinedFileRange : combinedFileRanges) {
+        CompletableFuture<ByteBuffer> result = new CompletableFuture<>();
+        combinedFileRange.setData(result);
+        boundedThreadPool.submit(() -> readCombinedRange(combinedFileRange, allocate));
+      }
+    }
+  }
+
+  /**
+   * function for reading combined or merged FileRanges. It reads the range and update the child
+   * fileRange's content.
+   *
+   * @param combinedFileRange merge file range, keeps track of source file ranges which were merged
+   * @param allocate Byte buffer allocator
+   */
+  private void readCombinedRange(
+      CombinedFileRange combinedFileRange, IntFunction<ByteBuffer> allocate) {
+    try (SeekableByteChannel channel = getReadChannel()) {
+      channel.position(combinedFileRange.getOffset());
+      ByteBuffer dst = allocate.apply(combinedFileRange.getLength());
```

Comment:
No, it's not a copy. Buffer is getting allocated for the range for the very first time here. We pass on the same buffer to gcsio layer, fill up the content and attach the same buffer to `CompletableFuture` of `FileRange` request.

In case of combined/merged ranges, we allocate single buffer based on the merging and then use `VectorReadUtil.slice` to provide different ByteBuffer views to child ranges. Underneath memory used will remain same and no extra copy is being performed.